src/runtime/typemanager.cs

```diff
@@ -484,11 +484,40 @@ internal static void InitializeSlots(IntPtr type, Type impl)
                     IntPtr slot = Interop.GetThunk(method);
                     Marshal.WriteIntPtr(type, offset, slot);
 
-                    seen[name] = 1;
+                    seen.Add(name);
                 }
 
                 impl = impl.BaseType;
             }
+
+            // Hack: for some objects (Classbase and ExtensionType) we need gc
+            // support. We need to provide three functions:
+            // - tp_traverse(object, ...) returns 0
+            // - tp_clear(object) returns 0
+            // - tp_is_gc(type) returns non-zero
+            // In addition, these functions can get called after a domain reload,
+            // so they must be implemented in native code.
+            //
+            // Py_GetVersion reliably returns non-zero and uses no arguments.
+            //
+            // PyAST_Check reads the object and returns zero unless it's a
+            // python AST node. You could violate this assumption by making a C#
+            // type that derives from ast.AST -- but it's not clear why you'd do that.
+            if (seen.Contains("tp_traverse"))
+            {
+                var lib = NativeMethods.LoadLibrary(Runtime.PythonDLL);
+                var zeroslot = NativeMethods.GetProcAddress(lib, "PyAST_Check");
+                var trueslot = NativeMethods.GetProcAddress(lib, "Py_GetVersion");
```

Comment:
What if the procs don't exist? I know it is unlikely, but should you catch an exception of some sort in that case?

Also, I think it would be good to make sure PyAST_Check returns 0 and Py_GetVersion returns non-zero. Is there a way you could add an assertion in Debug perhaps?